### PR TITLE
BlockHash endpoint implementation (retrieve)

### LINKED ISSUE #220: Get Block Hashes
ℹ️ This issue is part of Epic #202

## Objective
Implement the GET request: [Get Block Hashes](https://api.chainweb.com/openapi/#tag/blockhash/paths/~1chain~1%7Bchain%7D~1hash/get)

### PR DESCRIPTION
**Pull request by EdwinGuayacan**
Fixes #220 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [x] This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -805,6 +805,60 @@ Cut.retrieve(network_id: :mainnet01, location: "jp2", query_params: [maxheight:
...
+- `network_opts`: Network options. Keyword list with:
+  - `network_id` (required): Allowed values: `:testnet04` `:mainnet01`.
+  - `location` (optional): Location to access a Chainweb P2P bootstrap node. Allowed values:
+    - testnet: `"us1"`, `"us2"`, `"eu1"`, `"eu2"`, `"ap1"`, `"ap2"`
+    - mainnet: `"us-e1"`, `"us-e2"`, `"us-e3"`, `"us-w1"`, `"us-w2"`, `"us-w3"`, `"fr1"`, `"fr2"`, `"fr3"`, `"jp1"`, `"jp2"`, `"jp3"`
+  - `chain_id` (required): Id of the chain to which the request is sent. Allowed values: integer or string-encoded integer from 0 to 19.
+
+  - `query_params` (optional): Query parameters. Keyword list with:
+
+    - `limit` (optional): Integer (`>=0`) that respresent the maximum number of records that may be returned.
```
Comment (miguelnietoa) [inline README.md]: fix typo "respresent"

File: README.md [hunk 2]
```diff
@@ -805,6 +805,60 @@ Cut.retrieve(network_id: :mainnet01, location: "jp2", query_params: [maxheight:
...
+    - `next` (optional): String of the cursor for the next page. This value can be found as value of the next property of the previous page.
+    - `minheight` (optional): Integer (`>=0`) that represents the minimum block heigt of the returned headers.
+    - `maxheight` (optional): Integer (`>=0`) that represents the maximum block heigt of the returned headers.
```
Comment (miguelnietoa) [inline README.md]: fix typo "heigt"

File: lib/chainweb/p2p/block_hash.ex [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+  @endpoint "hash"
+
+  alias Kadena.Chainweb.P2P.BlockHashResponse
+  alias Kadena.Chainweb.{Error, Request}
+
+  @type network_opts :: Keyword.t()
+  @type error :: {:error, Error.t()}
+  @type block_hash_response :: BlockHashResponse.t()
+  @type retrieve_response :: block_hash_response | error()
```
Comment (miguelnietoa) [inline lib/chainweb/p2p/block_hash.ex]: IMO we dont need to define a type `block_hash_response`. Just keep as:
```elixir
@type retrieve_response :: BlockHashResponse.t() | error()
```

**Review by miguelnietoa — CHANGES_REQUESTED** [review]
@EdwinGuayacan @Odraxs good job guys!

**Review by miguelnietoa — APPROVED** [review]

File: lib/chainweb/p2p/block_hash.ex [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  @type network_opts :: Keyword.t()
+  @type error :: {:error, Error.t()}
+  @type retrieve_response :: BlockHashResponse.t() | error()
+
+  @spec retrieve(network_opts :: network_opts()) :: retrieve_response()
+  def retrieve(network_opts \\ []) do
+    location = Keyword.get(network_opts, :location)
+    network_id = Keyword.get(network_opts, :network_id, :testnet04)
+    chain_id = Keyword.get(network_opts, :chain_id, 0)
+    query_params = Keyword.get(network_opts, :query_params, [])
```
Comment (einerzg) [inline lib/chainweb/p2p/block_hash.ex:20]: I'm not sure but I don't see validation for this:

  `BlockHash.retrieve(query_params: [])`

Please correct me if I'm wrong.

**Review by einerzg — CHANGES_REQUESTED** [review]
@EdwinGuayacan  One comment!!!

File: lib/chainweb/p2p/block_hash.ex (hunk 4 above)
Comment (EdwinGuayacan) [reply lib/chainweb/p2p/block_hash.ex:20]: This is a valid use case, in which the default values are set and the API returns all items (4096).

**Review by einerzg — APPROVED** [review]

**Pull request merged**